### PR TITLE
Envoy 1.17 crashing with validation context secret discovery in ads mode on start up #15517

### PR DESCRIPTION
**Pull request by twghu**
Fix for [#15517](https://github.com/envoyproxy/envoy/issues/15517)

Commit Message: Prevent crash in case of ads configuration with no dynamic_resource.
Additional Description: See [Changes](https://github.com/envoyproxy/envoy/issues/15517#issuecomment-808203584)
Risk Level: TBD
Testing: Existing ADS test run, investigating creating custom one for this case.
Docs Changes: None
Release Notes: 
Dynamic Resource is required for ads configuration.  An ads configuration is not supported for primary clusters.
Platform Specific Features:

### CODE REVIEW

File: source/common/config/grpc_subscription_impl.cc [hunk 1]
```diff
@@ -29,6 +29,10 @@ void GrpcSubscriptionImpl::start(const std::set<std::string>& resources,
     init_fetch_timeout_timer_->enableTimer(init_fetch_timeout_);
   }
 
+  // An alternative to RELEASE_ASSERT here is to test for a null and generate warning error.
+  // This could potentially lead to an envoy instance with unintended configuration being 
+  // created with opaque security issues. It is probably best to fail.
+  RELEASE_ASSERT(grpc_mux_ != nullptr, "grpc_mux_ initialized with null value");
```
Comment (dmitri-d) [inline source/common/config/grpc_subscription_impl.cc:35]: I think the comment above isn't necessary, can't proceed here with `grpc_mux_` being nullptr.

File: source/common/config/subscription_factory_impl.cc [hunk 2]
```diff
@@ -86,8 +86,12 @@ SubscriptionPtr SubscriptionFactoryImpl::subscriptionFromConfigSource(
     }
   }
   case envoy::config::core::v3::ConfigSource::ConfigSourceSpecifierCase::kAds: {
+    Config::GrpcMuxSharedPtr gmuxsp = cm_.adsMux();
```
Comment (dmitri-d) [inline source/common/config/subscription_factory_impl.cc:89]: probably an `auto` here would suffice? Or skip the local var altogether, not sure if it improves readability.

File: source/common/config/subscription_factory_impl.cc [hunk 3]
```diff
@@ -86,8 +86,12 @@ SubscriptionPtr SubscriptionFactoryImpl::subscriptionFromConfigSource(
...
+    if ( gmuxsp == nullptr ) {
+      throw EnvoyException("A configuration for dynamic resources is required for ads with a non primary cluster");
```
Comment (dmitri-d) [inline source/common/config/subscription_factory_impl.cc:91]: Hmm, wasn't the issue here that sub-components (like SDS) of a primary cluster cannot be configured via ADS?

File: source/common/upstream/cluster_manager_impl.cc [hunk 4]
```diff
@@ -356,6 +355,9 @@ ClusterManagerImpl::ClusterManagerImpl(
           bootstrap.dynamic_resources().ads_config().set_node_on_first_message_only());
     }
   } else {
+    // if an ads is configured with no dynamic resource declared a nullptr leads to inconsistency in the
```
Comment (dmitri-d) [inline source/common/upstream/cluster_manager_impl.cc:358]: Not sure I understand this comment.

**Comment by dmitri-d** [comment]
Mostly nits, but could you add a test that verifies that a config similar to the one here: https://github.com/envoyproxy/envoy/issues/15517#issuecomment-805330213 generates an exception we expect?

**Comment by dmitri-d** [comment]
Please also check formatting (format check failed) and sign-off your commits (DCO checks isn't happy).

**Comment by lambdai** [comment]
When I was investigating, this issue is applied to main as well. I double checked with main branch.
Let's create the PR to main and backport to 1.17

File: source/common/config/subscription_factory_impl.cc (hunk 3 above)
Comment (lambdai) [reply source/common/config/subscription_factory_impl.cc:91]: Agree with @dmitri-d 
We could fix SDS, or explicitly disable SDS with primary cluster for now **as a workaround**.

Comment (dmitri-d) [reply source/common/config/subscription_factory_impl.cc:91]: We'd have to make SDS aware of the context in which it's being configured, which is a bit ugly; we could miss this issue in future components too if we handled this misconfiguration in SDS.

I'm not sure if it makes sense to continue launching Envoy if we hit this issue, baling with a very clear explanation about what the problem is looks like a good approach to me.

**Comment by twghu** [comment]
@lambdai Regarding test case.  Can you provide an example in existing code base that would best match "add the test case".  I've tried several options from existing code.  Tx.

**Comment by twghu** [comment]
I'll create a squashed commit for main branch and pr.

**Comment by twghu** [comment]
Closing as replaced with Main branch PR #15966

**Pull request closed without merging**